### PR TITLE
Fix: Rate/Deriv processing logic

### PR DESCRIPTION
**Pull request by Jiale-Fang**
# Description
- Consolidate range funcs with math funcs.
- Fix Metrics Range funcs
- Verify [Rate](https://prometheus.io/docs/prometheus/latest/querying/functions/#rate)/[Deriv](https://prometheus.io/docs/prometheus/latest/querying/functions/#deriv) processing logic
- Add irate()

# Testing
- "deriv(testmetric0[10m])"
- "rate(testmetric0[5m])"
- "irate(testmetric0[5m])"

# Checklist:
Before marking your pull request as ready for review, complete the following.

- [x] I have self-reviewed this PR.
- [x] I have removed all print-debugging and commented-out code that should not be merged.
- [x] I have added sufficient comments in my code, particularly in hard-to-understand areas.
- [x] I have formatted the code, if applicable. For Go, I have run `goimports -w .`.

### CODE REVIEW

File: pkg/integrations/prometheus/promql/metricsconsts.go
Comment (Jiale-Fang) [inline pkg/integrations/prometheus/promql/metricsconsts.go:1]: Add "isTimeRangeFunc" attr for prompt

File: pkg/segment/results/mresults/metricresults.go [hunk 1]
```diff
@@ -219,28 +218,31 @@ func (r *MetricsResult) AggregateResults(parallelism int) []error {
...
 	errList := []error{} // Thread-safe list of errors
 
+	// Use a temporary map to record the results modified by goroutines, thus resolving concurrency issues caused by modifying a map during iteration.
+	results := make(map[string]map[uint32]float64, len(r.Results))
+
 	var idx int
 	for grpID, timeSeries := range r.Results {
 		wg.Add(1)
-		go func(grp string, ts map[uint32]float64, function segutils.RangeFunctions) {
+		go func(grp string, ts map[uint32]float64, function structs.Function) {
```
Comment (nkunal) [inline pkg/segment/results/mresults/metricresults.go:235]: name this `ts` as `datapoints`

**Review by AndrewHess — APPROVED** [review]

**Comment by Jiale-Fang** [comment]
The rate here should actually be irate, I will modify it again and add the corresponding rate function.

**Comment by Jiale-Fang** [comment]
Implement rate and irate based on this: 
- https://techannotation.wordpress.com/2021/07/19/irate-vs-rate-whatre-they-telling-you/
- https://promlabs.com/blog/2021/01/29/how-exactly-does-promql-calculate-rates/

Implement derivative based on:
- https://en.wikipedia.org/wiki/Simple_linear_regression

**Review by AndrewHess — APPROVED** [review]

**Pull request merged**